### ISSUE TITLE
Unknown asm runtime version warning on startup

### ISSUE THREAD

**Issue by amir-khassaia** [eclipse/jetty.project#2412]
Jetty (9.4.9.v20180320) seems to be emitting an innocuous warning at startup:
`WARN AnnotationParser:90 - Unknown asm runtime version, assuming version 393216`

It ends up defaulting to the right version anyway but the warning is there nevertheless

ASM6 is oddly on the classpath yet `org.eclipse.jetty.annotations.AnnotationParser` is failing at

```
Package asm = Package.getPackage("org.objectweb.asm");
if (asm == null)
    LOG.warn("Unknown asm runtime version, assuming version {}", asmVersion);
```

Running under Java 8u131.

**Comment by joakime** [comment]
Did you recompress or rearchive or uber-ize or shade your jar files perchance?

**Comment by joakime** [comment]
Do you have a Java SecurityManager installed?
Are you using Java 9 (or newer) with JPMS?
Are you using OSGi?

**Comment by joakime** [comment]
## Testing results of `jetty-home-9.4.9.v20180320` 

Same `${jetty.base}` and `${jetty.home}`

``` none
[mybase]$ java -jar ../jetty-home-9.4.9.v20180320/start.jar --list-config | grep asm
16:                      6.0 | ${jetty.home}/lib/annotations/asm-6.0.jar
17:                      6.0 | ${jetty.home}/lib/annotations/asm-commons-6.0.jar
```

## with JDK 10

``` none
2018-04-04 17:11:07.925:INFO::main: Logging initialized @717ms to org.eclipse.jetty.util.log.StdErrLog
2018-04-04 17:11:08.355:INFO:oejs.Server:main: jetty-9.4.9.v20180320; built: 2018-03-20T07:21:10-05:00; git: 1f8159b1e4a42d3f79997021ea1609f2fbac6de5; jvm 10+46
2018-04-04 17:11:08.370:INFO:oejdp.ScanningAppProvider:main: Deployment monitor [file:///C:/code/jetty/distros/bases/websocket-threads/webapps/] at interval 1
2018-04-04 17:11:08.630:INFO:oeja.AnnotationConfiguration:main: Scanning elapsed time=33ms
2018-04-04 17:11:08.642:INFO:oejw.StandardDescriptorProcessor:main: NO JSP Support for /servlet25, did not find org.eclipse.jetty.jsp.JettyJspServlet
2018-04-04 17:11:08.724:INFO:oejs.session:main: DefaultSessionIdManager workerName=node0
2018-04-04 17:11:08.724:INFO:oejs.session:main: No SessionScavenger set, using defaults
2018-04-04 17:11:08.726:INFO:oejs.session:main: Scavenging every 660000ms
2018-04-04 17:11:08.755:INFO:oejsh.ContextHandler:main: Started o.e.j.w.WebAppContext@2d9caaeb{/servlet25,file:///C:/Users/joakim/AppData/Local/Temp/jetty-0.0.0.0-8080-servlet25.war-_servlet25-any-8929075017345302556.dir/webapp/,AVAILABLE}{C:\code\jetty\distros\bases\websocket-threads\webapps\servlet25.war}
2018-04-04 17:11:08.814:INFO:oejs.AbstractConnector:main: Started ServerConnector@461ad730{HTTP/1.1,[http/1.1]}{0.0.0.0:8080}
2018-04-04 17:11:08.814:INFO:oejs.Server:main: Started @1607ms
```

## with JDK 9

``` none
2018-04-04 22:12:32.713:INFO::main: Logging initialized @3193ms to org.eclipse.jetty.util.log.StdErrLog
2018-04-04 22:12:33.269:INFO:oejs.Server:main: jetty-9.4.9.v20180320; built: 2018-03-20T12:21:10Z; git: 1f8159b1e4a42d3f79997021ea1609f2fbac6de5; jvm 9+181
2018-04-04 22:12:33.287:INFO:oejdp.ScanningAppProvider:main: Deployment monitor [file:///mnt/c/code/jetty/distros/bases/websocket-threads/webapps/] at interval 1
2018-04-04 22:12:33.556:INFO:oeja.AnnotationConfiguration:main: Scanning elapsed time=27ms
2018-04-04 22:12:33.565:INFO:oejw.StandardDescriptorProcessor:main: NO JSP Support for /servlet25, did not find org.eclipse.jetty.jsp.JettyJspServlet
2018-04-04 22:12:33.668:INFO:oejs.session:main: DefaultSessionIdManager workerName=node0
2018-04-04 22:12:33.668:INFO:oejs.session:main: No SessionScavenger set, using defaults
2018-04-04 22:12:33.670:INFO:oejs.session:main: Scavenging every 600000ms
2018-04-04 22:12:33.706:INFO:oejsh.ContextHandler:main: Started o.e.j.w.WebAppContext@45ac5f9b{/servlet25,file:///tmp/jetty-0.0.0.0-8080-servlet25.war-_servlet25-any-11028382546500724856.dir/webapp/,AVAILABLE}{/servlet25.war}
2018-04-04 22:12:33.801:INFO:oejs.AbstractConnector:main: Started ServerConnector@4a07d605{HTTP/1.1,[http/1.1]}{0.0.0.0:8080}
2018-04-04 22:12:33.801:INFO:oejs.Server:main: Started @4282ms
```

No WARN emitted.

**Comment by amir-khassaia** [comment]
Thanks for confirming with your output.
This is under Java 8, no JPMS, no OSGI, no shaded/uber Jars, no security manager.
Embedded Jetty being started with standard classpath and jetty/asm Jars in a folder on the classpath. Also occurs when starting from Intellij and gradle managed classpath.

Embedded context configurations are slightly customised but I dont see anything that should impact that (eg AnnotationConfiguration mostly calls through to super which ends up trying to locate asm by default and doesnt): 
```
 context.setConfigurations(new Configuration[]{
                new WebInfConfiguration() ,
                new WebXmlConfiguration(),
                new MetaInfConfiguration(),
                new FragmentConfiguration(),
                new AnnotationConfiguration() {
                    @Override
                    public void preConfigure(final WebAppContext context) throws Exception {
                        super.preConfigure(context);

                        if (config.isRunningInDevelopmentEnvironment()) {
                            if (!devClassesPresentInClasspath()) {
                                enableDevWebInitializerWorkaround(context);
                            }
                        }
                    }                    

                    private void enableDevWebInitializerWorkaround(final WebAppContext context) {
                        Set<String> value = ConcurrentHashMap.newKeySet();
                        value.add(WebAppInitializer.class.getName());
                        // hard-coded class names for runtime dependencies
                        value.add("some-web-initializer-reference");

                        ClassInheritanceMap map = new ClassInheritanceMap();
                        map.put("org.springframework.web.WebApplicationInitializer", value);
                        context.setAttribute(CLASS_INHERITANCE_MAP, map);
                    }

                    @Override
                    public void configure(final WebAppContext context) throws Exception {
                        logger.debug("Starting annotation configuration...");
                        final Stopwatch stopWatch = Stopwatch.createStarted();
                        super.configure(context);
                        logger.debug("Finished annotation configuration in {}ms.", stopWatch.elapsed(MILLISECONDS));
                    }
                },
                new JettyWebXmlConfiguration()
        });
```

**Comment by joakime** [comment]
I'm suspecting you have more than 1 ASM present in your classpath.

Run this in your project and check ...

``` java
package cloader;

import java.net.URL;
import java.util.Enumeration;

public class WhereIsThisComingFrom
{
    public static void main(String[] args)
    {
        locate(org.objectweb.asm.Opcodes.class);
    }

    public static void locate(Class<?> clazz)
    {
        try
        {
            ClassLoader cl = clazz.getClassLoader();
            String classAsResource = clazz.getName().replace('.', '/') + ".class";
            Enumeration<URL> urls = cl.getResources(classAsResource);
            System.out.printf("Looking for: %s%n", classAsResource);
            while (urls.hasMoreElements())
            {
                URL url = urls.nextElement();
                System.out.printf("Found: %s%n", url.toExternalForm());
            }
        }
        catch (Throwable t)
        {
            System.out.printf("Whoops: cannot locate: %s%n", clazz);
            t.printStackTrace();
        }
    }
}
```

**Comment by joakime** [comment]
Also, if you are going to enable the `AnnotationConfiguration`, know that the `org.springframework.web.SpringServletContainerInitializer` will be found and used, you are essentially doing the initialization twice with that setup.

**Comment by amir-khassaia** [comment]
I've checked the jar list in the deployed app, and in dev gradle dependencies output and IDE classpath listing, all show single ASM6.

It seems to be more indicative of classloader not being aware of asm at all which doesnt quite make sense.

I did run your sample program too which also shows one.
```
"C:\Program Files\Java\jdk1.8.0_161\bin\java" -Dfile.encoding=UTF-8 -classpath C:\Users\amir\AppData\Local\Temp\classpath1124147422.jar biz.papercut.pcng.server.web.WhereIsThisComingFrom
Looking for: org/objectweb/asm/Opcodes.class
Found: jar:file:/C:/Users/amir/.gradle/caches/modules-2/files-2.1/org.ow2.asm/asm/6.0/bc6fa6b19424bb9592fe43bbc20178f92d403105/asm-6.0.jar!/org/objectweb/asm/Opcodes.class

Process finished with exit code 0

```

Also thanks for the tip on initializer, it was only needed in a specific instance in dev environment on windows due classpath size limit and having intellij work around it by using a dynamic classpath as per the command line output above too, in that specific case the custom WebInitializer was not on the list and needed to be declared up front to be found.

**Comment by amir-khassaia** [comment]
FYI: gradle dependencies snippet, no other ASM dependency is present elsewhere
```
+--- org.eclipse.jetty:jetty-annotations:9.4.9.v20180320
|    +--- org.eclipse.jetty:jetty-plus:9.4.9.v20180320
|    |    +--- org.eclipse.jetty:jetty-webapp:9.4.9.v20180320
|    |    |    +--- org.eclipse.jetty:jetty-xml:9.4.9.v20180320
|    |    |    |    \--- org.eclipse.jetty:jetty-util:9.4.9.v20180320
|    |    |    \--- org.eclipse.jetty:jetty-servlet:9.4.9.v20180320
|    |    |         \--- org.eclipse.jetty:jetty-security:9.4.9.v20180320
|    |    |              \--- org.eclipse.jetty:jetty-server:9.4.9.v20180320
|    |    |                   +--- javax.servlet:javax.servlet-api:3.1.0
|    |    |                   +--- org.eclipse.jetty:jetty-http:9.4.9.v20180320
|    |    |                   |    +--- org.eclipse.jetty:jetty-util:9.4.9.v20180320
|    |    |                   |    \--- org.eclipse.jetty:jetty-io:9.4.9.v20180320
|    |    |                   |         \--- org.eclipse.jetty:jetty-util:9.4.9.v20180320
|    |    |                   \--- org.eclipse.jetty:jetty-io:9.4.9.v20180320 (*)
|    |    \--- org.eclipse.jetty:jetty-jndi:9.4.9.v20180320
|    |         \--- org.eclipse.jetty:jetty-util:9.4.9.v20180320
|    +--- org.eclipse.jetty:jetty-webapp:9.4.9.v20180320 (*)
|    +--- javax.annotation:javax.annotation-api:1.2
|    +--- org.ow2.asm:asm:6.0
|    \--- org.ow2.asm:asm-commons:6.0
|         \--- org.ow2.asm:asm-tree:6.0
|              \--- org.ow2.asm:asm:6.0
```

**Comment by amir-khassaia** [comment]
Update from my end, I have a workaround: explicitly loading any asm class before starting AnnotationConfiguration - as it has a dependency on asm being loaded for its check to work.

I dont quite follow why it occurs in my setup only - there must be something nuanced... but `java.lang.Package#getPackage(java.lang.String)` may return null if the current ClassLoader had not yet been made aware of the class.
So this flow is not guaranteed to be stable as there is no insurance that in the context of the ClassLoader used by the caller of the AnnotationParser the asm classes have been loaded (at this point in time).

Specifically getPackage states:
> Find a package by name in the callers ClassLoader instance

In my case I can make it work by loading the class into the classloader:
Ie. Calling either: `org.objectweb.asm.Opcodes.class.getName();` or    `Class.forName("org.objectweb.asm.Opcodes");` before calling super, eg:

 ```
context.setConfigurations(new Configuration[]{
                new AnnotationConfiguration() {
                    @Override
                    public void configure(final WebAppContext context) throws Exception {
                        Class.forName("org.objectweb.asm.Opcodes");
                        super.configure(context);
                    }
                }

```

**Comment by janbartel** [comment]
@amir-khassaia how do you define the classpath at runtime? Also, what are the classes of the classloaders in use?

**Comment by janbartel** [comment]
Also, @amir-khassaia can you double-check that you don't have 2 copies of the jetty-annotations jar on the classpath: one on the container path and one on the webapp classpath? Have you done any setup of system and server classes at all, if so please post here also.

Please enable server dump after start, then do a run with all debug turned on (org.eclipse.jetty.LEVEL=DEBUG) and post here.

Finally, I don't see how its possible for the classloader not to have loaded an asm class in the static initializer of AnnotationParser, because of there is this static field, which refers to an asm class:
  
```
protected static int ASM_OPCODE_VERSION = Opcodes.ASM6;
```

**Comment by amir-khassaia** [comment]
The classpath is defined using standard Java SE semantics, there are standard classloaders in use: 
Classpath scan config: 
`org.eclipse.jetty.server.webapp.ContainerIncludeJarPattern set to ".*/*\\.jar$|.*/classes/.*"`

Thats a good point about the static constant that should be bringing the class in but the nuance here is that the static constant in AnnotationParser is initialized first however it may not trigger class loading as its referencing an implicit constant, eg similar to https://stackoverflow.com/a/22650503
Doing this `new Opcodes() {};` instead of referencing constants or any of the other actions from earlier message will load it...

**Comment by janbartel** [comment]
@amir-khassaia can you provide more information on your jvm and operating system?

I've been testing this on various flavours of jdk8 on linux and can't reproduce it. This is the code (derived from the ServerWithAnnotations class in the jetty embedded/exmples source tree).  I've been running it with mvn exec:java -Dexec.mainClass="org.eclipse.jetty.embedded.ServerWithAnnotations".  Can you run it in your environment and post  the results (ideally you'll run it against the same standard jetty  test webapp from the distro, but change to yours if necessary).


``` java
 import java.io.File;

import org.eclipse.jetty.plus.jndi.EnvEntry;
import org.eclipse.jetty.plus.jndi.Resource;
import org.eclipse.jetty.plus.jndi.Transaction;
import org.eclipse.jetty.security.HashLoginService;
import org.eclipse.jetty.server.Server;
import org.eclipse.jetty.webapp.Configuration;
import org.eclipse.jetty.webapp.FragmentConfiguration;
import org.eclipse.jetty.webapp.JettyWebXmlConfiguration;
import org.eclipse.jetty.webapp.MetaInfConfiguration;
import org.eclipse.jetty.webapp.WebAppContext;
import org.eclipse.jetty.webapp.WebInfConfiguration;
import org.eclipse.jetty.webapp.WebXmlConfiguration;
import org.eclipse.jetty.annotations.AnnotationConfiguration;

/**
 * ServerWithAnnotations
 */
public class ServerWithAnnotations
{
    public static final void main( String args[] ) throws Exception
    {
        // Create the server
        Server server = new Server(8080);

        // Create a WebApp
        WebAppContext webapp = new WebAppContext();
        webapp.setConfigurations(
                new Configuration[]{                        
                    new WebInfConfiguration() ,
                    new WebXmlConfiguration(),
                    new MetaInfConfiguration(),
                    new FragmentConfiguration(),
                    new AnnotationConfiguration() 
                    {
                        @Override
                        public void preConfigure(final WebAppContext context) throws Exception 
                        {
                            super.preConfigure(context);                
                        }

                        @Override
                        public void configure(final WebAppContext context) throws Exception 
                        {
                            System.err.println("Starting annotation configuration...");
                            super.configure(context);
                            System.err.println("Finished annotation configuration.");
                        }
                    },
                    new JettyWebXmlConfiguration()});
        webapp.setContextPath("/");
        File warFile = new File(
                "../../jetty-distribution/target/distribution/demo-base/webapps/test.war");
        webapp.setWar(warFile.getAbsolutePath());
        webapp.setAttribute(
                "org.eclipse.jetty.server.webapp.ContainerIncludeJarPattern",
                ".*/javax.servlet-[^/]*\\.jar$|.*/servlet-api-[^/]*\\.jar$");
        server.setHandler(webapp);

        // Configure a LoginService
        HashLoginService loginService = new HashLoginService();
        loginService.setName("Test Realm");
        loginService.setConfig("src/test/resources/realm.properties");
        server.addBean(loginService);

        server.start();
        server.join();
    }

}
```

**Comment by amir-khassaia** [comment]
Environment: Windows 10, JDK 8u161.

I will try example + standard jetty project later and get back to you.
It might be beneficial to run both yours and mine with verbose class loading as well (`-verbose:class`) to have confirmation of the class loading order for the class in question.

I'm also including debug log as I haven't done so as per your earlier request in case it reveals another angle.

[jetty-start-up.log](https://github.com/eclipse/jetty.project/files/1878741/jetty-start-up.log)

**Comment by janbartel** [comment]
@amir-khassaia  what is ```file:///C:/Users/amir/AppData/Local/Temp/classpath870036071.jar```

Does that jar contain all of the classes of your dependencies merged into 1 big jar, or does it have a manifest that specifies a classpath - if that latter, please provide the manifest.

Also, I think you're really going to want to change that containerIncludeJarPattern to narrow it down  - really no point scanning all the jre classes!?

**Comment by amir-khassaia** [comment]
Yep, sorry for the confusion on the include jar pattern I relaxed it only as part of this investigation.

The classpath jar contains the manifest that lists the classpath references inside. I've seen it outside dev environment too where dynamic classpath via manifest is not used so I dont think its a factor.

I've uploaded zipped copy of the dynamic classpath jar anyway.
[classpath467371086.zip](https://github.com/eclipse/jetty.project/files/1878848/classpath467371086.zip)

**Comment by janbartel** [comment]
@amir-khassaia I just noticed something which may be relevant. The log warning is the same whether we can't load the Package, or we can't get the specification version (see https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-annotations/src/main/java/org/eclipse/jetty/annotations/AnnotationParser.java#L90  and also https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-annotations/src/main/java/org/eclipse/jetty/annotations/AnnotationParser.java#L95).

Can you double check that your asm jar has a manifest that lists the Implementation-Version header? I just updated to asm-6.1.1 and it does *not* specifiy the Implementation-Version!

Meanwhile, I will update the code to ensure that the 2 error messages are distinct.

**Comment by janbartel** [comment]
@amir-khassaia I've made the change to the log message, and also (to eliminate doubt) changed the code to explicitly load the Opcode class. The changes are in the branch "jetty-9.4.x-issue-2397-jdk10-support". That branch also updates to asm-6.1.1. If you can test the code out, that would be great.

**Comment by amir-khassaia** [comment]
@janbartel Yes I noticed the logging issue too but its not a factor for me as I debugged through to confirm it was the first one strictly involving locating ASM class. ASM6 distribution pulled in from maven repos contains full manifest with impl version, I've checked that too.

Nevertheless I've looked over your commits and they look good to me as they should sort this out.

I will need to have a go at building your branch before I can test it out.

**Comment by amir-khassaia** [comment]
Having issues building the branch, but I can confirm the commit changes fix the issue if applied locally. Ie. specifically the `Opcodes.class.getPackage()` invocation.

Happy to try a snapshot build if you have one handy ?

What would be the timeline for a release with this change ?

**Comment by janbartel** [comment]
We have another couple of changes that we need to get into a release, so I expect that there will be another release in the next couple of weeks.

**Comment by janbartel** [comment]
Closing.
